### DISCUSSION TITLE
Max mandatory overmap specials and mods

### DISCUSSION THREAD

**Discussion by b3brodie** [CleverRaven/Cataclysm-DDA#47036] [Q&A] [answered]
With the map special rebalance form #46701 I have found that my existing mod list now throws the too many mandatory overmap specials error from overmap.cpp on line 426 "There are too many mandatory overmap specials (%d > %d). Some of them may not be placed".  From what I can find the maximum count of mandatory specials appears to be calculated with two different frequencies.  Due to this, does the presence of the error mean that all of the mandatory specials are just less likely to be found, or does it mean that there is a chance that they simply do not spawn in the world at all?

**Comment by actual-nh** [comment]
What is your mod list, BTW?

Ping: @Venera3

**Reply by b3brodie** [reply]
I have a large mod list totaling 83 mandatory specials, but you can achieve this error using only fewer farms, aftershock, crit expansion, and magiclysm.  If I recall correctly all of these come with the base game.

Edit: Fewer farms doesn't come with the game but you can get close to the limit using just the other 3 mods.

**Reply by anothersimulacrum** [reply]
Fewer farms is third party.

**Comment by anothersimulacrum** [comment]
> Due to this, does the presence of the error mean that all of the mandatory specials are just less likely to be found, or does it mean that there is a chance that they simply do not spawn in the world at all?

There's technically a chance that a special can not spawn at all if it is not mandatory, but it's incredibly unlikely (if it has proper placement requirements). When there are too many mandatory specials, the mandatory ones will not be culled, so they will still have opportunities to be placed.

**Reply by b3brodie** [reply]
Alright, thank you for the information

**Comment by Venera3** [comment]
All optional specials are suppressed until all mandatory  specials are placed, so your overmaps are gonna look pretty samey.
The weird thing is, currently there are *no* mandatory specials in the base game. The only way I could imagine something getting borked is one of the mods overwriting the flags of a special set to UNIQUE without touching their occurrences. If you start up the game and one special dominates (you might want to spin up some debug games for this) that'll be the problem.

**Reply by b3brodie** [reply]
I looked through some of the mods on my list and I believe at least a good portion of the mandatory specials was due to fewer farms.  In the description it states that it converts farms into specials, and while I am not sure the exact amount that this converts it is at least 12 based on the lack of errors that results from removing it from the list.

**Reply by anothersimulacrum** [reply]
Consider making a bug report on their github repo?

**Reply by b3brodie** [reply]
I put an issue in there but from my understanding of how the mod works I'm not sure if it can be fixed.  I'm just going to remove it to stop having these errors.

**Comment by kevingranade** [comment, marked as answer]
This has been resolved by #47071 
The consistency check was based on an old version of how specials worked, i.e. "if there is a minimum value, it's mandatory", which has since been changed just for specials with the UNIQUE flag, which never count as mandatory unless the min and max value are the same, for example if a mod author overloads the game with mandatory specials by setting them all to "occurrences": [ 1, 1 ], and then tries to get the warning to go away by setting "flags": [ "UNIQUE" ].

Also, if you look at the PR, it essentially obsoletes the fewer farms mod by doing the same thing, greatly reducing the occurrence of farms.
